spmd/compiler/api.py

```diff
@@ -298,9 +317,24 @@ def _convert_to_distributed(
             assert i < len(
                 inps
             ), f"got more placeholer nodes ({i + 1}) than inputs ({len(inps)})"
-            node_to_obj[node] = DTensor.from_local(
-                inps[i], schemas[i].mesh, schemas[i].placements
-            )
+            if training_phase == TrainingPhase.FORWARD:
+                # in the forward phase we start with the full "global" ensors.
+                # we needed this because we needed to capture the original graph.
+                node_to_obj[node] = distribute_tensor(  # DTensor.from_local(
+                    inps[i],
+                    schemas[i].mesh,
+                    schemas[i].placements,
+                )
+            else:
+                # But in the backward pass we got "real" sharded inputs
+                # so we have to actually make DTensors out of them
+                assert training_phase == TrainingPhase.BACKWARD
```

Comment:
I didn't fully get this. I thought for both forward and backward, we first get graphs of local tensors and expand that. Did the plan change?